File: src/main/java/core/driver/DriverManager.java
package core.driver;

import static core.utils.CapabilitiesHelper.readAndMakeCapabilities;

import org.openqa.selenium.remote.DesiredCapabilities;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;

import constants.Target;
import exceptions.PlatformNotSupportException;
import io.appium.java_client.AppiumDriver;

public class DriverManager {
    private static AppiumDriver driver;
    // For Appium < 2.0, append /wd/hub to the APPIUM_SERVER_URL
    final String APPIUM_SERVER_URL = "http://127.0.0.1:4723";

    public AppiumDriver getInstance(Target target) throws IOException {
        System.out.println("Getting instance of: " + target.name());
        return switch (target) {
            case ANDROID -> getAndroidDriver();
            case IOS -> getIOSDriver();
        };
    }

    private AppiumDriver getAndroidDriver() throws IOException {
        var map = readAndMakeCapabilities("android-caps.json");
        return getDriver(map);
    }

    private AppiumDriver getIOSDriver() throws IOException {
        HashMap map = readAndMakeCapabilities("ios-caps.json");
        return getDriver(map);
    }

    private AppiumDriver getDriver(HashMap map) {
        DesiredCapabilities desiredCapabilities = new DesiredCapabilities(map);

        try {
            driver = new AppiumDriver(
                    new URL(APPIUM_SERVER_URL), desiredCapabilities);
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }

        return driver;
    }
}

File: src/main/java/core/driver/DriverManager.kt
package core.driver

import constants.Target
import io.appium.java_client.AppiumDriver
import org.openqa.selenium.WebElement
import org.openqa.selenium.remote.DesiredCapabilities
import java.net.URI

class DriverManager {
    fun getInstance(target: Target): AppiumDriver<WebElement>? {
        println("Getting instance of: " + target.name)
        return when (target) {
            Target.ANDROID -> getAndroidDriver()
            Target.IOS -> getIOSDriver()
        }
    }

    private fun getAndroidDriver(): AppiumDriver<WebElement>? = getDriver(
        mapOf(
            "platformName" to "android",
            "automationName" to "uiautomator2",
            "platformVersion" to "15",
            "deviceName" to "Automation",
            "app" to "src/main/resources/app-debug.apk",
            "appPackage" to "io.appium.android.apis",
            "appActivity" to "io.appium.android.apis.ApiDemos"
        )
    )

    private fun getIOSDriver(): AppiumDriver<WebElement>? = getDriver(
        mapOf(
            "platformName" to "iOS",
            "automationName" to "XCUITest",
            "deviceName" to "iPhone 13",
            "app" to "src/test/resources/TestApp.app"
        )
    )

    private fun getDriver(map: Map<String, *>): AppiumDriver<WebElement>? {
        val url = URI("http://127.0.0.1:4723").toURL()
        return AppiumDriver<WebElement>(url, DesiredCapabilities(map))
    }
}


